### PR TITLE
[FIRRTL] Replace RHS Sinks used as Sources with Wires

### PR DESCRIPTION
**Pull request by seldridge**
Change the FIRRTL parser to remove usages of sinks as _right hand side_ (RHS) expressions (either output ports or instance inputs that show up on the RHS). These are replaced with an intermediary wire and all connections are updated.

Note: this is currently partially handled by casts using `firrtl.asPassive`. However, this PR pushes towards limiting `asPassive` casts to only things which have in-out behavior. Namely, registers and wires. If anything else is used illegally in an in-out way (e.g., a module output port or an instance input port), then it is replaced by a wire and the wire is cast `asPassive` when needed.

There are a lot of examples in the tests, but a simple example is below. Here, port `a` is illegally used as a source.

```
circuit Foo :
  module Foo :
    output a : UInt<1>
    output b : UInt<1>
    a is invalid
    b <= a
```

This is converted to:

```
circuit Foo :
  module Foo :
    output a : UInt<1>
    output b : UInt<1>
    wire a_synthetic : UInt<1>
    a <= a_synthetic
    a_synthetic is invalid
    b <= a_synthetic
```

This PR adds the ability to compute Scala FIRRTL Compiler "flow", but doesn't use it. This is merely a convenience for computing if something needs to have an intermediary wire added.

### Relationship to Chisel3 emitted FIRRTL IR ("CHIRRTL")

Chisel3 allows users to use almost anything as a RHS, including things which are traditionally "sinks". I.e., you can use an L-value as an R-value freely. The two common examples include:

1. Reading from an output port
2. Reading from an instance input port

While weird, this is actually quite useful in two situations:

1. You have no idea how one of these L-values was computed, but you want to read it's value. E.g., the value is being set inside some deep `when`/`.elsewhen`/`.otherwise` structure.
2. You want to "tap" some signal with a monitor or an assertion.

A full example could look like (for an interactive version see: https://scastie.scala-lang.org/DUXfJREnRee7nGA8CkgSgQ):

```scala
class Foo extends MultiIOModule {
  val a = IO(Output(Bool()))
  val sel = IO(Input(Bool()))
  
  a := 0.U
  when (sel) {
    a := 1.U
  }
  
  experimental.verification.assert(a =/= 0.U)
}
```

This produces the following. Note that `a` is used as an R-value when computing the node associated with the assert:

```
circuit Foo :
  module Foo :
    input clock : Clock
    input reset : UInt<1>
    output a : UInt<1>
    input sel : UInt<1>
    a <= UInt<1>("h0")
    when sel :
      a <= UInt<1>("h1")
    node _T = neq(a, UInt<1>("h0"))
    assert(clock, _T, UInt<1>("h1"), "")
```

Problematically, this results in the FIRRTL IR that Chsiel emits (usually called "CHIRRTL") to be illegal FIRRTL IR as defined in the FIRRTL specification. You aren't supposed to be able to use "sinks" as "sources". Because this is used liberally in Chisel code, sinks are allowed to be used as sources and never removed. Consequently, ports almost behave like wires (they have in-out semantics). This PR tries to remove that usage and relegate in-out semantics to only wires and registes.

With this PR, the above code example is immediately cleaned up to add a synthetic wire in place of the problematic `a`. This then relegates "wire"-like behavior to wires and registers and preserves the correct semantics of ports (that they are true R or L-values).

### CODE REVIEW

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 1]
```diff
@@ -20,11 +20,15 @@
...
 #include "mlir/Translation.h"
 #include "llvm/ADT/ScopedHashTable.h"
+#include "llvm/ADT/SetVector.h"
 #include "llvm/ADT/SmallPtrSet.h"
 #include "llvm/ADT/StringExtras.h"
+#include "llvm/ADT/TypeSwitch.h"
 #include "llvm/Support/SourceMgr.h"
 #include "llvm/Support/raw_ostream.h"
 
+#include <iostream>
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:30]: I know this is a WIP, but please don't use iostream.  You can use `llvm::errs() <<` instead of cerr.  The issues are explained in the llvm coding standards.

Comment (seldridge) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:30]: Cool! This is what I was looking for (to printf debug what was going on), but couldn't find it. :+1:

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 2]
```diff
@@ -56,6 +58,102 @@ static StringAttr filterUselessName(StringAttr name) {
   return isUselessName(name.getValue()) ? StringAttr() : name;
 }
 
+/// Return a nullable Value that is, optionally, the port that this
+/// value originates from.
+static void getPorts(Value value, SmallVector<Value, 4> &ports) {
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:63]: Small suggestion:
```suggestion
static void getPorts(Value value, SmallVectorImpl<Value> &ports) {
```

**Comment by seldridge** [comment]
There are intermittent failures here for the following circuit. @darthscsi indicated that this is likely due to the iteration order in the `SmallSetVector rhsSinks` and I'm replacing things in (sometimes) the wrong order):

```
circuit Baz:
  module Foo:
    input a: Clock
    input b: Clock

  module Baz:
    input a: Clock
    output b: Clock

    inst foo of Foo

    b <= foo.a
    b <= foo.b

    foo.a <= a
    foo.b <= a
```

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp (hunk 2 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp]: Cool! Didn't know this, but it makes sense... I've included this in the rebase.

**Comment by seldridge** [comment]
@darthscsi: The failing test case you identified is now fixed. I opted to just not erase the dead operands. These will get cleaned up eventually (and do get cleaned up with something like `firtool`).

This should be ready for a review.

Note: this is currently rebased on top of #416 and includes work (converting more arguments, clocks, resets, etc. to passive) which could be factored into other, prerequisite PRs or direct pushes.

**Comment by youngar** [comment]
<details>
<summary>This test case is problematic</summary>

```
circuit A :
  module A :
    input i : UInt<1>
    output o : UInt<1>
    o <= i
  module B :
    input i : UInt<1>
    output o : UInt<1>
    inst a of A
    a.i <= i
    o <= a.o
    node problem = and(a.i, a.o)
```

```
Case1.fir:12:20: error: 'firrtl.and' op operand #1 must be sint or uint type, but got '!firrtl.flip<uint<1>>'
    node problem = and(a.i, a.o)
```

```
module  {
  firrtl.circuit "A" {
    firrtl.module @A(%i: !firrtl.uint<1>, %o: !firrtl.flip<uint<1>>) {
      firrtl.connect %o, %i : !firrtl.flip<uint<1>>, !firrtl.uint<1>
    }
    firrtl.module @B(%i: !firrtl.uint<1>, %o: !firrtl.flip<uint<1>>) {
      %0 = firrtl.wire : !firrtl.flip<uint<1>>
      %a = firrtl.instance @A {name = "a"} : !firrtl.bundle<i: flip<uint<1>>, o: uint<1>>
      %1 = firrtl.subfield %a("i") : (!firrtl.bundle<i: flip<uint<1>>, o: uint<1>>) -> !firrtl.flip<uint<1>>
      firrtl.connect %0, %i : !firrtl.flip<uint<1>>, !firrtl.uint<1>
      %2 = firrtl.subfield %a("o") : (!firrtl.bundle<i: flip<uint<1>>, o: uint<1>>) -> !firrtl.uint<1>
      firrtl.connect %o, %0 : !firrtl.flip<uint<1>>, !firrtl.flip<uint<1>>
      %3 = firrtl.subfield %a("i") : (!firrtl.bundle<i: flip<uint<1>>, o: uint<1>>) -> !firrtl.flip<uint<1>>
      firrtl.connect %3, %0 : !firrtl.flip<uint<1>>, !firrtl.flip<uint<1>>
      %4 = firrtl.asPassive %0 : (!firrtl.flip<uint<1>>) -> !firrtl.uint<1>
      %5 = firrtl.subfield %a("o") : (!firrtl.bundle<i: flip<uint<1>>, o: uint<1>>) -> !firrtl.uint<1>
      %6 = firrtl.and %4, %0 : (!firrtl.uint<1>, !firrtl.flip<uint<1>>) -> !firrtl.uint<1>
      %problem = firrtl.node %6 {name = "problem"} : !firrtl.uint<1>
    }
  }
}
```
</details>

**Comment by seldridge** [comment]
Nice find, @youngar. Thanks for the concise test case. Taking a look...

**Comment by seldridge** [comment]
@youngar: I added that test and have it passing in the rebase. I was accidentally replacing all references to the instance instead of just the bad subfield (`%6 = firrtl.and %4, %0` shouldn't be `%0`). My tests are somewhat light on the instance side, so I'm not ruling out other issues.

**Comment by seldridge** [comment]
There are a few corner cases that weren't considered related to dealing with partial connects that have no common types. I need some representation better than `nullptr` or need to special case that.

Example crashing circuit:

```
circuit Foo :
  module Foo :
    output a: {a: UInt<1>}
    output b: UInt<1>
    a is invalid
    b <- a
```

**Comment by lattner** [comment]
Hey @seldridge I'm sorry i'm just now getting a chance to look at this.

My understanding of this patch is that you're trying to simplify passes that work on the FIRRTL Dialect by getting it closer to the design intent of the FIRRTL specification, fencing off weirdness that we have to accept by making the parser accept it but rewrite it away.  This is nice because nothing in the MFC compiler will have to deal with outputs that are read from etc.

Have you considered a different approach here?  Instead of embracing the (violated) design intent of the written FIRRTL spec, why don't we just embrace the design reality of the empirical FIRRTL spec (as defined by "what is actually output by Chisel") and just treat flipped values as being `inout`?

The consequence of this is that a naive impl of LowerToRTL would turn each flipped port into an `!rtl.inout` port instead of an output result.  We can improve on that naive implementation by doing some dataflow analyis in LowerToRTL or by just doing a naive lowering and do a promotion pass as an RTL to RTL pass.

The downside of this is that we'd have to treat flipped values a having inout semantics in the FIRRTL passes, but I think that's already effectively what the discussions about flow are pointing to.

WDYT?

**Comment by seldridge** [comment]
I agree with your analysis here.

After thinking about this, what this PR is really doing is pushing the promotion of "weird RHS sinks" being converted to inout from `LowerToRTL` into `FIRParser`. This, coupled with adding an `inout` type in the FIRRTL dialect that wraps wires and registers, would enable verification of connect statements.

Unfortunately, I seem to only have just discovered that `LowerToRTL` is doing the same work that this PR is doing (just later)... 😬 

On the `main` branch, if you start with the following circuit (where the output `%b` drives itself):

```mlir
firrtl.module @Foo(%a: !firrtl.uint<1>, %b: !firrtl.flip<uint<1>>) {
  firrtl.connect %b, %a : !firrtl.flip<uint<1>>, !firrtl.uint<1>
  firrtl.connect %b, %b : !firrtl.flip<uint<1>>, !firrtl.flip<uint<1>>
}
```

Running `-lower-to-rtl` will remove the weird RHS sink usage of `%b` by promoting it to a wire:

```mlir
rtl.module @Foo(%a: i1) -> (%b: i1) {
  %0 = rtl.wire : !rtl.inout<i1>
  rtl.connect %0, %a : i1
  %1 = rtl.read_inout %0 : !rtl.inout<i1>
  rtl.connect %0, %1 : i1
  rtl.output %1 : i1
}
```

With this PR, this promotion happens earlier and the above is parsed into:

```mlir
firrtl.module @Foo(%a: !firrtl.uint<1>, %b: !firrtl.flip<uint<1>>) {
  %b_synthetic = firrtl.wire : !firrtl.flip<uint<1>>
  firrtl.connect %b, %b_synthetic : !firrtl.flip<uint<1>>, !firrtl.flip<uint<1>>
  firrtl.connect %b_synthetic, %a : !firrtl.flip<uint<1>>, !firrtl.uint<1>
  firrtl.connect %b_synthetic, %b_synthetic : !firrtl.flip<uint<1>>, !firrtl.flip<uint<1>>
}
```

Which gets lowered into the following (notably this is the same as running on `main` except that `%0 -> %b_synthetic`):

```mlir
rtl.module @Foo(%a: i1) -> (%b: i1) {
  %b_synthetic = rtl.wire : !rtl.inout<i1>
  rtl.connect %b_synthetic, %a : i1
  %0 = rtl.read_inout %b_synthetic : !rtl.inout<i1>
  rtl.connect %b_synthetic, %0 : i1
  rtl.output %0 : i1
}
```

So, rephrasing your question: does promotion to `inout` happen during parsing or do we delay it and fix it during RTL conversion?

I would love if we could verify FIRRTL dialect connect statements directly in that dialect as opposed to relying on lowering to RTL. The current problem is that we can't differentiate between any of the following because the FIRRTL dialect type system is lacking:

```mlir
firrtl.module @Foo(%a: !firrtl.uint<1>, %b: !firrtl.flip<uint<1>>) {
  // This is legal if %a is a register, but illegal for ports
  firrtl.connect %a, %a : !firrtl.uint<1>, !firrtl.uint<1>
  // This is legal if %a and %b are wires or registers, but illegal for ports
  firrtl.connect %a, %b : !firrtl.uint<1>, !firrtl.flip<uint<1>>
  // This is legal unconditionally
  firrtl.connect %b, %a : !firrtl.flip<uint<1>>, !firrtl.uint<1>
  // This is legal if %b is a register
  firrtl.connect %b, %b : !firrtl.flip<uint<1>>, !firrtl.flip<uint<1>>
}
```

This is, as you identify, going to make writing passes over FIRRTL dialect tricky because the safeguards and type/flow/kind checking (which we arguably could have in the FIRRTL dialect with the addition of an `inout` type and no need for kind or flow) that the SFC provides aren't possible to enforce in the FIRRTL dialect (you can write almost anything in the FIRRTL dialect currently). If we go that route, then any new pass work should probably happen on the RTL dialect and we should discourage passes on the FIRRTL dialect. We then accept the fact that the FIRRTL dialect can't support the same types of checks as the SFC.

So, I think the question is do we want to have FIRRTL dialect as a first-class dialect or is it something that is narrowly used for getting things converted to RTL and the real work, checking, cool passes happen there?

**Comment by lattner** [comment]
Thanks @seldridge I agree with your analysis.  I think you've correctly identified the crux of the issue:

> So, I think the question is do we want to have FIRRTL dialect as a first-class dialect or is it something that is narrowly used for getting things converted to RTL and the real work, checking, cool passes happen there?

I'd personally prefer the later: use the FIRRTL dialect as the thinnest layer of lowering we can do to get from Chisel to RTL, then do the heavy lifting at the RTL level.  This is for three major reasons:

1) FIRRTL dialect is awkward and bug prone to deal with, as a consequence of being the strata in which width inference is implemented (e.g. lots of redundant ops like head/tail, need to defensively guard against unknown widths, implicit promotions and wonky width rules, issues with flow, annotations instead of parameterization, etc) 

2) I'm not aware of anything the design points above "makes easier".  We need to build out a parameterized RTL system etc, but once we do, I don't see any rationale for doing heavy lifting at the FIRRTL level.  I don't think we necessarily even need to do inlining at the FIRRTL level.

3) The RTL level will be shared by other frontends, so work there is going to be much higher leverage on the community.

If you agree with this, then the major purpose of the FIRRTL dialect is to support FIRRTL lowering: when elimination, width inference, bundle lowering, etc.  I think we can ground the conversation on "what is necessary to enable those things".

-Chris

**Comment by seldridge** [comment]
@clattner wrote:
> If you agree with this, then the major purpose of the FIRRTL dialect is to support FIRRTL lowering: when elimination, width inference, bundle lowering, etc. I think we can ground the conversation on "what is necessary to enable those things".

Yes. I agree with this. We can clarify this in the FIRRTL rationale doc.

I'll close this, but plan to re-use some pieces of this as needed. Specifically the utilities involving flow and computing the common type between two bundles will help with lowering partial connects. I'll break these out on an as-need basis.

**Comment by lattner** [comment]
Sounds great, thanks Schuyler!  Agreed on updating the rationale doc too, thx
